### ISSUE TITLE
VGGImageClassifier does not use ReLU in Head and doesn't support "flatten" as pooling.

### ISSUE THREAD

**Issue by bermeitinger-b** [keras-team/keras-hub#2616]
Yes, we are still experimenting with VGG, and we're wondering why certain models and datasets perform so poorly in a few of our experiments.

We are instantiating the VGGImageClassifier, e.g., from the present `vgg_11_imagenet`. The problem is that the classification head does not use ReLU activation functions by default. It uses a linear one also when instantiated from the presets. The head looks like this for a classification task with 10 classes:
```
Conv2D(4096) → Conv2D(4096) → Pooling → Dense(10) 
```
All the activations are linear, not just the last one. However, shouldn't the Conv2D activations be ReLU?

The constructor has `activation=None` as the default, but when it is set to `activation="relu"`, all the above layers use ReLU, including the final output layer. The last one should be linear (or softmax).

I would also like to point out a crash when using the original paper idea: `Flatten()→Dense(4069)→Dense(4069)→Dense(10)`. This should be possible by setting `pooling="flatten"`, but then `self.pooler` is a double-head configuration, which crashes because the shapes do not match anymore:
```
Conv2D(4096) → Conv2D(4096) → Flatten() → Dense(4096)  → Dense(4096)  → Dense(10) 
```

You can see the experiments in this notebook:
https://colab.research.google.com/drive/1lPGsOxCLaB6usbrlDeiCrU2V1FA0w0e8?usp=sharing

I have a suggestion on how to fix, but I'm unsure if this would have any side-effects with loading presets or converting weights: https://github.com/keras-team/keras-hub/compare/master...bermeitinger-b:keras-hub:fix-vgg?expand=1

**Comment by maitry63** [comment]
Hi @bermeitinger-b , 
You are correct  in the original VGG paper, `ReLU` is used after each Dense layer, but Keras defaults to linear. This isn’t a change to VGG itself, it’s just how Keras‑Hub standardizes classifier heads across multiple backbones.

In VGGImageClassifier, the classifier head uses linear Dense layers by default. This comes from a generic builder used for backbones like ResNet, EfficientNet, and VGG.

When no activation is specified, it defaults to linear.  Using linear helps keep gradients flowing smoothly during training. ReLU can sometimes cause "dead neurons" (stuck at zero) in deep heads, especially with large 4096-unit layers.

For original VGG behavior like you mentioned we can customize it:

```
x = keras.layers.Dense(4096, activation="relu")(x)
x = keras.layers.Dense(4096, activation="relu")(x)
x = keras.layers.Dense(num_classes, activation="softmax")(x)
```


While using pooling="flatten" directly often fails because input_shape is sometimes ignored by Keras‑Hub, and the backbone outputs dynamic spatial dimensions (None, None, None, 512). Flatten cannot determine the input size for the Dense layer.

Means:
When we do, `pooling = "flatten"` 

The classifier does: `Flatten() to Dense(4096)`. 
But because the spatial dimensions are None, Keras tries to build a Dense layer with: shape = (None, 4096), causing an error.

we can try to wrap the backbone in a Functional model with a fixed input shape so Keras can compute static spatial dimensions before building the Dense layers.

Please refer this [gist](https://colab.sandbox.google.com/gist/maitry63/b0839923fe1286cfe7f7339eab5d9cf7/vggimageclassifier-does-not-use-relu-in-head-and-doesn-t-support-flatten-as-pooling-2616.ipynb).

Let me know if this helps!

**Comment by bermeitinger-b** [comment]
Thank you for your response. I know how to implement VGG correctly. However, I still believe this is a mistake in keras_hub. I understand that ReLU is a quickly saturated activation function, but it is still used in the original paper, and if the class is called _VGG_, I should be able to trust that it actually is _VGG_.

Look at the code:
https://github.com/keras-team/keras-hub/blob/9569150618c5db4252587ab45cf2724fd5b0b874/keras_hub/src/models/vgg/vgg_image_classifier.py#L134-L142

This pooler is included in the final head, constructing a non-working (and nonsensical) model:
https://github.com/keras-team/keras-hub/blob/9569150618c5db4252587ab45cf2724fd5b0b874/keras_hub/src/models/vgg/vgg_image_classifier.py#L149-L180


Your argument also contradicts the docstring:
https://github.com/keras-team/keras-hub/blob/9569150618c5db4252587ab45cf2724fd5b0b874/keras_hub/src/models/vgg/vgg_image_classifier.py#L41-L43

This is wrong. This `activation` is also set in the Conv2D/Dense layers, where it doesn't make sense to use _softmax_. The docstring is about the latest layer, the classification layer. There, it makes sense to use softmax, linear, or even sigmoid, but not ReLU. (Additionally, it actually doesn't default to `softmax,` but that's another issue.

**Comment by maitry63** [comment]
@bermeitinger-b , Thank you reporting this issue. 
You’re correct , the current [VGGImageClassifier](https://keras.io/keras_hub/api/models/vgg/vgg_image_classifier/) has a docstring and head-design inconsistency.

The docstring states that the `activation` argument applies only to the Dense layer and defaults to `softmax `(which is not actually required), but in reality it also applies to the `Conv2D` layers in the head, and the default is not `softmax`.

The head structure `(Conv2D → pooler → Dense)` does not match the original VGG fully-connected layers, which can be confusing.

Also, using `pooling="flatten"` directly (as described in the docstring) can fail because the backbone outputs dynamic spatial dimensions `(None, None, None, 512)`, so Flatten cannot infer the input size.

This is indeed a bug. We will update the documentation to clarify, which layers the activation applies to, the correct default, and the distinction between intermediate and final activations.

**Comment by bermeitinger-b** [comment]
Thank you. However, I'm not confused. I'm reporting an issue with the VGG implementation. If you actually follow the code, setting `pooling="flatten"` will never work and is nonsensical. Please look at the code. This has nothing to do with the shapes but with control flow.

**Comment by maitry63** [comment]
@bermeitinger-b , 
Yes the issue with` pooling="flatten"` is not just a shape problem [ refer this [gist.](https://colab.sandbox.google.com/gist/maitry63/b0839923fe1286cfe7f7339eab5d9cf7/vggimageclassifier-does-not-use-relu-in-head-and-doesn-t-support-flatten-as-pooling-2616.ipynb#scrollTo=oK25479VwaDP) 
In the current VGGImageClassifier implementation, does not match the original VGG fully-connected layers.

As a result, setting `pooling="flatten"` will never produce a valid VGG head, regardless of input shape. This is a implementation bug, not merely a documentation or dynamic shape issue.

We will look into it throughly.

Thank you bringing this up!
